### ISSUE TITLE
ndsctl sometime is not reacting to commands

### ISSUE THREAD

**Issue by sqrt-1764** [nodogsplash/nodogsplash#387]
I am running OpenWrt 18.0.6 on a Fritzbox 4020
NDS is Version 3.2.1 as offered by OpenWrt repository

Occasionally I ndsctl seems to get stuck.  

    ndsctl status

as well as  

    ndsctl stop

do not react any more. I have to stop the commands with ^C

In the background NDS seems to be working. Some log entries indicate normal background behaviour.

    pgrep nodogsplash

also returns a process id so the process is still running.

But when I execute the above commands, I get

    daemon.err nodogsplash[17820]: Socket is not ready for communication : Invalid argument
    daemon.err nodogsplash[17820]: Socket is not ready for communication : Bad file descriptor

in the log.
Seems to happen when the box is busy. Otherwise it runs without problems for some days.

Stopping and restarting the service brings ndsctl back to life again.

**Comment by bluewavenet** [comment]
@sqrt-1764 
I have encountered similar symptoms before, on a 32MB RAM device. It was caused by the device running out of free RAM as it was also running a web server.

I have also seen the problem on a dual radio ath10k device with 128MB RAM, the 5GHz drivers seem to use a huge amount of memory when busy with lots of connections.

Your Fritzbox 4020 has 128MB RAM and is only single radio so I think so running out is unlikely unless you are running something else that is using the free memory. 
What does `free` report?

**Comment by sqrt-1764** [comment]
It's only a snapshot of the current state (restarted NDS 4h ago):  

                 total       used       free     shared    buffers     cached
    Mem:        125272      37412      87860        964       2212      13004
    -/+ buffers/cache:      22196     103076
    Swap:            0          0          0

(My webserver is nginx with php-fpm - to complete the technical details)

So currently plenty of free space.  
This might change when a request to NDS is processed (resulting in a sub process querying ndsctl and some curl actions to log some state) But I doubt that memory should be the problem here as there were less than 4 devices in the WLAN when ndsctl showed this behavior.

Would a very long redirect-url (meaning 400 characters or so ...) be a problem for NDS? I observerd some when this problem occured.

**Comment by bluewavenet** [comment]
@sqrt-1764 
Are you using FAS or a splash meta redirect?

FAS restricts the query string length to 2K. If it is longer NDS just truncates it.
However the redirect url (ie the url the client used to trigger CPD is usually very short. eg:
 http://captive.apple.com/hotspot-detect.html
 http://connectivitycheck.android.com/generate_204
being by far the most common.

It is of course hardly worth returning these or even your own landing page as more and more, individual CPD implementations are now immediately closing the CPD browser as soon as an Internet connection is detected - the detection is done by a rapid polling of the detection URL in the background, for example Firefox does this every 3 seconds.

In summary, if you are getting unusual or long query strings then something is wrong.

**Comment by sqrt-1764** [comment]
I am using FAS - but the strings should be far shorter than 2k
I only mentioned it in case NDS isn't hardened against that case - which might occur. See below.

Well, that is not the case.
I observed some Apple-Devices that queried some iTunes updates or something like this. This over http before the device was authenticated, so NDS kicked in and served the splash-page to them. The redirect-url occupied more than 2 lines in my terminal ...
And this happens automatically without user interaction - so there is a good chance that a user might not react to CPD anyway.

iPhones still leave the redirect open after authentication ... As do desktops. But with Android-Phones you are right. So one can not rely on the fact that this page is displayed.

But as we now redirect to a constant (our ;-) ) page, I will remove that redirect url in the future ...

**Comment by sqrt-1764** [comment]
For the reference:  
Here a freshly recorded redir link (as it is encrypted I see no privacy issues in posting this here)  
http://iosapps.itunes.apple.com/itunes-assets/Purple113/v4/05/6a/55/056a5562-f6cd-f404-86ca-fb4020611d82/mzps.1727109713522508255.ipa?accessKey=1563601357_3153055203714759563_Gh9oMeFDZGrEeMgufiLe7ef7CS8Y6J3Agy26Zlv5ULsLYvI+lrTDFhceAS7LNevRwbGO3HbTDU9sBAKbDcaDPj2MD3SndxU0S3H/fmC8PVtTyCz488gnFqwCo93C44VXbgOm3dVJLmnyYZYL/PnMpy7y6bx7M4AmpdNBPETAD2I9cxB0d194JfhtCI0b0hE+

**Comment by bluewavenet** [comment]
That length should not be a problem at all.
However is is very unusual for a client device to do this, the CPD will only use its fixed detection url.

On a very busy site I have searched the FAS server logs for iosapps.itunes.apple.com over the last 8000 separate logins (about 8 weeks worth) and there are no instances of this. 
However, http://captive.apple.com/hotspot-detect.html occurs over 2000 times.

What type of site is this? My example is a hotel.

You could add logging of the user agent string to your FAS. This might give us some insight.

None of this should effect ndsctl though. If you have  lot of client streaming/downloading perhaps you have a cpu issue? 

On my example site NDS is running on a PCengines APU2, but it has run for a while on a Gl-iNet B1300 (similar to your Fritzbox 4020) while I upgraded the version of OpenWrt on the APU2. No issues seen on either.

**Comment by sqrt-1764** [comment]
The site is a bar. Usage currently is minimal (10 users a day with traffic < 100MB each). Only mail or whatsapp, some browsing ... resulting in sessions with not much traffic. So with my hardware I expect no bottlenecks at the moment.  

Those URLs were recorded in the access-log of the nginx on the router (that only is used for NDS) so it only serves unauthenticated http-requests at the router.  

But I think I can rule out those URLs.


Yesterday NDS stopped again. Again no clues in the log-files, what might be happening. This time there were no long URLs in the vicinity of the "crash".  
Again "ndsctl" was responding, but "ndsctl stop" and "ndsctl status" were not responding.  "Free" reported the values from above.  
As I am not on site, I can not see the current behaviour at the client devices, but yesterday I got a screenshot that showed nginx with a 502-Error page.
So I also looket at the error log from nginx. The error was reported at 22:50 (where I was already sleeping) and the corresponding entries there are:  

'
2019/07/19 20:39:58 [error] 1374#0: *5340 upstream timed out (145: Unknown error) while reading response header from upstream, client: 192.168.101.159, server: , request: "GET /nodog/fas.php?clientip=192.168.10$
2019/07/19 20:40:58 [error] 1374#0: *5340 upstream timed out (145: Unknown error) while reading response header from upstream, client: 192.168.101.159, server: , request: "GET /nodog/fas.php?clientip=192.168.10$
2019/07/19 22:48:15 [error] 1374#0: *5357 upstream timed out (145: Unknown error) while reading response header from upstream, client: 192.168.101.178, server: , request: "GET /nodog/fas.php?clientip=192.168.10$
2019/07/19 22:49:15 [error] 1374#0: *5357 upstream timed out (145: Unknown error) while reading response header from upstream, client: 192.168.101.178, server: , request: "GET /nodog/fas.php?clientip=192.168.10$
2019/07/19 22:51:32 [error] 1374#0: *5362 upstream timed out (145: Unknown error) while reading response header from upstream, client: 192.168.101.159, server: , request: "GET /nodog/fas.php?clientip=192.168.10$
2019/07/19 22:51:39 [error] 1374#0: *5365 no live upstreams while connecting to upstream, client: 192.168.101.159, server: , request: "GET /nodog/fas.php?clientip=192.168.101.159&gatewayname=gn-101-01&redir=htt$
2019/07/19 22:51:39 [error] 1374#0: *5366 no live upstreams while connecting to upstream, client: 192.168.101.159, server: , request: "GET /nodog/fas.php?clientip=192.168.101.159&gatewayname=gn-101-01&redir=htt$
2019/07/19 22:51:39 [error] 1374#0: *5367 no live upstreams while connecting to upstream, client: 192.168.101.159, server: , request: "GET /nodog/fas.php?clientip=192.168.101.159&gatewayname=gn-101-01&redir=htt$
2019/07/19 22:53:43 [error] 1374#0: *5370 upstream timed out (145: Unknown error) while reading response header from upstream, client: 192.168.101.159, server: , request: "GET /nodog/fas.php?clientip=192.168.10$
2019/07/19 22:55:12 [error] 1374#0: *5373 upstream timed out (145: Unknown error) while reading response header from upstream, client: 192.168.101.159, server: , request: "GET /nodog/fas.php?clientip=192.168.10$
2019/07/19 22:56:12 [error] 1374#0: *5373 upstream timed out (145: Unknown error) while reading response header from upstream, client: 192.168.101.159, server: , request: "GET /nodog/fas.php?clientip=192.168.10$
2019/07/19 23:07:25 [error] 1374#0: *5376 upstream timed out (145: Unknown error) while reading response header from upstream, client: 192.168.101.159, server: , request: "GET /nodog/fas.php?clientip=192.168.10$
2019/07/19 23:08:25 [error] 1374#0: *5376 upstream timed out (145: Unknown error) while reading response header from upstream, client: 192.168.101.159, server: , request: "GET /nodog/fas.php?clientip=192.168.10$
'
As you can see I got those errors also before, but only some of them with no reports of failures from the useres (whatever that means ;-) ) From hostapd.log it seems that the WLAN-Connection was stable for that client at the moment.

My current question is:  
What causes ndsctl to stop responding only(!) to "ndsctl stop" and "ndsctl start"? The answer to this question might lead to the source of the error - which then can be fixed.
Currently I think that the hardware is perfectly capable of handling the tasks and my configuration also is correct.  
I organized my code so that the majority of the splash page can be handled locally from the router. So I have a nginx with php-fpm running that queries ndsctl locally, queries a central server with curl for dynamic content and also logs to that server via curl.
From my splash page (written in PHP) I call ndsctl with
'
exec("sudo /usr/bin/ndsctl json ".$this->clientip." 2>&1", $clientData, $retCode);
'
which seems fine to me. For logging then I have some calls with curl to my central server which might be a source of problems in case the connection is bad. 
But then I would expect my splash page to crash, not NDS.

**Comment by sqrt-1764** [comment]
... one more question:  
Is there any documentation what the log-levels mean? Curently I am running at the defalut level of 5.  
Would it make sense to increase the log-level?

I configured OpenWrt to log to the central server, so size of the logfile doesn't matter. ;-)

**Comment by mwarning** [comment]
@sqrt-1764 you can try 10, higher does not make sense. But anyway, we should make that less confusing.

**Comment by bluewavenet** [comment]
@sqrt-1764
In your logs, the get requests seem to be truncated with a $ symbol....

Try setting loglevel to 7, then look at logread output when a client preauthenticates.
It should report "splashpageurl:" and this is what we need to check when you get the error.

**Comment by sqrt-1764** [comment]
Well - the log comes from nginx, not NDS, so changing the loglevel would not change anything there.

But according to @mwarning I now increased the loglevel to the maximum of 10.
Looking into the logs, I already can see entries starting with "splashpageurl" there ...

But currently no new crash from NDS - i will report, when I get a new one.

**Comment by bluewavenet** [comment]
@sqrt-1764
While waiting for a crash, show me one of the "working" examples of splashpageurl please.

**Comment by sqrt-1764** [comment]
This logentry corresponds to a successful login afterwards:  
`2019-07-20T13:13:03+01:00 gn-101-01 nodogsplash[1516]: splashpageurl: http://192.168.101.1:2051/nodog/fas.php?clientip=192.168.101.132&gatewayname=gn-101-01&redir=http%3a%2f%2fcaptive.apple.com%2fhotspot-detect.html'

**Comment by sqrt-1764** [comment]
@bluewavenet: Maybe you have more insight. I also observed one case where a Huawei P20 Pro assiciated to the WLAN. I checked nodogsplash.log and found
```
2019-07-20T13:36:59+01:00 gn-101-01 nodogsplash[1516]: originurl: http://connectivitycheck.platform.hicloud.com/generate_204_0226130d-9ffa-41cc-beb1-d312be1e29b4
2019-07-20T13:36:59+01:00 gn-101-01 nodogsplash[1516]: splashpageurl: http://192.168.101.1:2051/nodog/fas.php?clientip=192.168.101.154&gatewayname=gn-101-01&redir=http%3a%2f%2fconnectivitycheck.platform.hicloud.com%2fgenerate_204_0226130d-9ffa-41cc-beb1-d312be1e29b4
```

But strangely no access to nginx was logged.  
There the connection was very short and maybe the client alredy disconnected.
Or is there a chance that this URL made problems to NDS and this was a lost opportunity to display the splash page?

**Comment by bluewavenet** [comment]
Splashpageurl is of course where NDS does its 307 redirect.
This is what your nginx should see as the GET request.  When you have the error reported in ngnix, the GET request seems to be truncated with a $ symbol. I am trying to see if it is NDS sending this in error or if something is going wrong elsewhere.

**Comment by bluewavenet** [comment]
@sqrt-1764
Your Huawei P20 Pro log looks quite normal and correct.
I think your ngnix is only logging errors?

**Comment by sqrt-1764** [comment]
Not clear what you are meaning.  
splashpageurl => 307 is clear to me. So far everything looks ok there.
My nginx is logging access as well as errors (in separate logs). (This is another reason why I changed from uhttp to nginx) The logentry might be truncated - but only the entry (did not change anything at the configuration of nginx there, but a truncated log entry would make sense to me. I think I remember such a statement in the docs)  

What bothers me is that in case of the Huawei no GET-Request was generated after the response from NDS - otherwise it would have been reported in the access log .
This would mean that either the Huawei did not respond to the 307 (but here the timing is so tight ...) or that NDS did not output the 307. This is why I asked.

**Comment by bluewavenet** [comment]
@sqrt-1764
I have checked in the logs of my large site for "connectivitycheck.platform.hicloud.com" and there are 191 successful logins triggered by this out of the last 4K. But this was running initially NDS v3.3.2 and later v4.0.0. There have been many changes since v3.2.1 so perhaps you have uncovered a bug that has since been fixed albeit without knowing....

**Comment by sqrt-1764** [comment]
I will monitor this further. Maybe it was a glitch. Currently we have not very much traffic on our portal ...  

Vers. 3.2.1 is the current one on OpenWrt for my device and the current Version is almost always the better choice. I would be willing to update to the current version. Is there any place to find instructions how to update to 4.0 on OpenWrt? The instructions here in the wiki seem a little bit old (referencing to OpenWrt 14.4 if i see right).

Off-Topic: An option to get the MAC from the device in the 307 request would be handy for me. It would save the call to ndsctl from my FAS-Script if the token also can be transmitted securely with the 4.0 (when I read right) ... Would be another argument to invest in the Version 4.0 ;-)

Off-Topic (2): I now am wondering which version "stable" on readthedocs is referring to. On the first glance I can not determine if it is 4.x or 3.x ... As the comments in the stable branch here already mention 4.0 i might read the wrong version of the documentation?

**Comment by bluewavenet** [comment]
V4.0.0 was only released last week so stable and latest documentation are currently identical.

I will test 4.0.0 on openwrt 18 and if good I will show you how to upgrade.

The binaries are available on openwrt snapshot and will be in openwrt 19 when it comes out.

https://nodogsplashdocs.readthedocs.io/en/stable/fas.html

**Comment by sqrt-1764** [comment]
Looking forward to 4.0 ;-)

Hmm, I do not remember "fas_secure_enable" to be present in the documentation when I read it some 2 or 3 moths ago. So I assume this does not apply to the 3.2.1 that is currently deployed on OpenWrt. Would be a source for headaces for new users, i think ...

**Comment by bluewavenet** [comment]
It does apply to 3.2.1 but fas_secure_enable=1
is the default :-D

**Comment by sqrt-1764** [comment]
Ähem ...  
I now re-read the documentation and found some parts that were not present some weeks ago. "fasremotefqdn" I read for the first time. (I also tried to run FAS an my central server until I noticed that it would be difficult to access ndsctl on the router from there to get the MAC of the device. If this option had been mentioned there I wouldn't have overlooked it, i think.)  
Maybe I was blind (but I doubt so) but the documentation mentions in the first chapter of the FAQ that FAS Secure Level 2 was introduced by v4. So I should not try to use it in 3.2.1, right?

**Comment by bluewavenet** [comment]
"Stable" documentation always refers to the current release (4.0.0 at the moment).
"Latest" documentation always refers to Master on Github. As 4.0.0 was only released last week, and no changes have been merged, Latest and Stable are at this point in time identical.

You are correct, FAS Secure Level 2 was introduced in v4, as well as"fasremotefqdn".

Version 4 is specifically enhanced to make running a FAS on a centralised server an easy task.
I have it running on a custom OpenWrt 18.x image for the x86-64 architecture so I am pretty sure t will be ok on any 18.x, but I have dot done the tests.
I can try on a standard 18.06.4/ar71xx tomorrow and report back, but feel free to try it yourself if you have a test system.
http://downloads.openwrt.org/snapshots/packages/mips_24kc/routing/nodogsplash_4.0.0-1_mips_24kc.ipk

**Comment by sqrt-1764** [comment]
For the records:  
Still on 3.2.1. I now have observed another device, that did not get the splash page:

Connected to the WLAN at 13:49:54 (-> hostapd.log)  
DHCPACK at 13:49:57
2019-07-21T13:49:58+01:00 gn-101-01 nodogsplash[1516]: splashpageurl: http://192.168.101.1:2051/nodog/fas.php?clientip=192.168.101.147&gatewayname=gn-101-01&redir=http%3a%2f%2fconnectivitycheck.gstatic.com%2fgenerate_204
But no corresponding entry in the access.log from nginx.  
This time the device is still connected 10 minutes later (as reported by iwinfo) with no disconnect in the logs.

**Comment by sqrt-1764** [comment]
Regarding the documentation:
IMHO the versions should be mentioned more clearly. It also should be made more clear which option was introduced when. Currently an OpenWrt User gets wrong information if he does not read the documentation completely (e.g. it should be mentioned that FAS Secure Level is only valid for v4 where that option is described, not only hidden in the FAQ ...) Or even better - provide a vesion clearly marked with v3.2.x as long as this version is distributed by OpenWrt.

**Comment by bluewavenet** [comment]
@sqrt-1764
Re: Documentation - Good feedback thanks.
I'll see what can be done!

**Comment by bluewavenet** [comment]
@sqrt-1764 
> 2019-07-21T13:49:58+01:00 gn-101-01 nodogsplash[1516]: splashpageurl: http://192.168.101.1:2051/nodog/fas.php?clientip=192.168.101.147&gatewayname=gn-101-01&redir=http%3a%2f%2fconnectivitycheck.gstatic.com%2fgenerate_204

This log seems normal and correct. I would suspect that if NDS reports this it will also have served the 307 redirect, otherwise these would be an error logged just after this.
My guess is that the client CPD did not open its browser window. This can happen when the device blanks its screen or turns off its wifi to save battery (depending on how it is configured).

I don't think this is anything to do with the ndsctl hanging issue though.

Has the ndsctl issue occurred again yet?

**Comment by bluewavenet** [comment]
@sqrt-1764 
As promised I have tested v4.0.0 on OpenWrt 18.06.4 (nodogsplash_4.0.0-1_mips_24kc.ipk) and as expected it works fine.

**Comment by sqrt-1764** [comment]
1) No "crash" yet. It occured every 2 or 3 days in the past ...

2) I will now also test the 4.0. First on my FritzBox 4040 where I develop.  
As I intend to move to FAS Secure 2 there will be some adjustments made to the code ...

**Comment by sqrt-1764** [comment]
Regarding the log:  
As I understand it, the client gets presented a 307 page from NDS when this entry in the nodogsplash-log is generated. So my question is, what could block the client from requesting the page from requesting that page? As I am not on site, I can not ask the user. ;-)  
One option would be, that the signal got bad then and the request did not make it through. As that DHCP request did not prodouce a network name, I do not know what device it was ...

I do all this not (only) because of the crashes, but also because of the problems of the splash page not being displayed. This is solved for iPhones now, but maybe there are other devices ...

**Comment by bluewavenet** [comment]
@sqrt-1764 
Once installed you will find a fully functional "example" php script (fas-aes.php) in /etc/nodogsplash.
This should just work if copied to your local ngnix server and specified in faspath.
For a remote FAS, also set fasremoteip and fasremotefqdn.

Details are in the default config file as well as the documentation.
Note:
  * The "php-cli" package and the "php-openssl" module must both be installed for fas_secure level 2.
  * Nodogsplash does not depend on this package and module, but will exit gracefully
   if this package and module are not installed when this level is set.

**Comment by bluewavenet** [comment]
> .... because of the problems of the splash page not being displayed

This can happen as a consequence of how some CPD implementations work.

If the device has used the NDS captive portal SSID recently (and remembers it), but has been deauthed by NDS (authidletimeout or checkinterval expired) AND can also see another SSID that it also knows, it may try the captive portal, sees the 307 redirect, but instead of opening its CPD browser, immediately tries the other SSID it knows. In this way your FAS is not touched.

The CPD is reinitialised by turning WiFi off and on, on most devices so only occurs if the device has gone out of range long enough to time out but has kept its WiFi on.

Increasing authidletimeout is a good solution. The amount depends on the type of site....

**Comment by sqrt-1764** [comment]
Grrr! ;-)
The Fritzbox 4040 uses another processor. Apparently it is arm_cortex-a7_neon-vfpv, so I looked in
https://downloads.openwrt.org/snapshots/packages/arm_cortex-a7_neon-vfpv4/packages/
for an .ipk but unfortunately found none.

**Comment by sqrt-1764** [comment]
... found it in "routing" ;-)

**Comment by sqrt-1764** [comment]
Peace of cake ;-))
Installed 4.0 on my two 4020 as well as my 4040 without problems. As the configuration file was compatible I did not have to make any changes yet.

With one exception:  
Now the max. value of loglevel is 7 otherwise NDS won't start. Something for the documentation.

**Comment by bluewavenet** [comment]
To be honest I thought loglevel 7 always was the max... never mind ;-)

**Comment by sqrt-1764** [comment]
@mwarning stated obbove that 10 was the max value - so I had that value in my configuration file ...

**Comment by bluewavenet** [comment]
https://nodogsplashdocs.readthedocs.io/en/stable/debug.html

**Comment by bluewavenet** [comment]
I wonder why 3.2.1 will start with loglevel=10 and 4.0.0 will not?
I have a vague recollection from years ago that higher than 7 always gave an error message, but ....

**Comment by bluewavenet** [comment]
From the config file:
```
  # Enables debug output (0-7)
  #option debuglevel '7'
```
But ndsctl allows you to set "loglevel" to any level.
I have almost always used ndsctl to switch it on and off without restarting so not noticed this difference before.

**Comment by sqrt-1764** [comment]
I can confirm that this was also present in the original config from 3.2.1  
But when told to set it to 10 i did so -  and 3.2.1 did not complain whereas 4.0 did.

**Comment by bluewavenet** [comment]
Weird. The code that checks dates back at least to v3.0.0.
A diversion from the issue at hand, but interesting.

**Comment by sqrt-1764** [comment]
Ah, now I remember:  
I set the loglevel via ndsctl. Also changed the config file but never rebooted.
So v4.0 was the first that read that config-file ... now everything comes together ;-)

**Comment by mwarning** [comment]
The loglevel should be is limited to 7. But, in reality it is platform dependent. It needs to be fixed.

**Comment by sqrt-1764** [comment]
After I upgraded to NDS v4.0 everything ran smooth.
Now, after about 4 days, I ndsctl got stuck again.
No suspicious entries in the system-log.

But I noticed something strange. Maybe there is some useful information in the following:
First some remarks about my FAS page to understand my observations in my logfile. My FAS page does the following:
- get the time and the GET/POST parameter and store them in memory
- call ndsctl for the MAC from the ip-address (I haven't migrated to FAS Secure 2 yet)
- make a curl-request to my central server for usage data for that device/user
- output the html
- finally get the system time again and log the recorded data to my central server

Today had very little usage, so the log messages were easy to review.
At 19:33:08 I had an access that was served in 230 ms with a redir-url in the GET-request of ```"http://captive.apple.com/hotspot-detect.html"```  
At 19:33:21 two overlapping requests from the same device that were served in 350ms each. Redir-URL from the GET-Request were
```"http://iosapps.itunes.apple.com/itunes-assets/Purple113/v4/0b/31/69/0b3169fd-7995-5e93-5f75-562d1858510b/mzps.5922383458986037994.ipa?accessKey=1564272820_1521543316314281882_jAH69cnSou+p82kbfuFvCwg1OnX/ea/L/rkknbLu6lessoWi54tyUGJRxFati2/z+uCVgpkhYHB8kdfVG84G0loHw9FfwNRqxO6SejdJWtvqr6bBD8lAgQYHcoZ+qHwkxMdZ69EOr+hsm1HQG9ScjcPLl95hKn6ekT10gARkB6ucSEGQMFevTpzHSk9E0Arz" ``` 
and  
```"http://iosapps.itunes.apple.com/itunes-assets/Purple123/v4/6a/63/4a/6a634a2c-6422-6aba-2a08-bca0eba9232c/mzps.3227220458391780027.ipa?accessKey=1564272819_8519936701553187034_lEDtqMcOnnnvZh6JyIPfP+swETVW70zpuE6LeXEIlf/mrGPIo0bmLOdPw0IGUeuJcn4XwOiCZUMtmQSDo+0bNF19Qwp373XvLk32qiRDgaZY7lZhwB61zb+KT0JrNXI1B6DlHBBHbPJvdtXe3clP9+U1mQd85L5PyYkXKLsNKjdxUhrP0ijVUgVVhqsC/cae"```

The next request at 19:33:22 from that device (again with such a long redir-url) now was served in 2.05s  
The following at 19:33:25 in 2.3s (-> ```http://captive.apple.com/hotspot-detect.html```)

Another device at 19:35:46 was served in 53s (-> ```ttp://captive.apple.com/hotspot-detect.html```)
followed by 60,3s at 19:37:15.

That was about the time when I noticed that ndsctl did not respond any more to ```ndsctl status```
free showed plenty of free memory (like above), top showed no cpu-usage and ps showed nodogsplash running.
After I restarted nodogsplash with ```/etc/init.d/nodogsplash stop``` and restarted accordingly, ndsclt worked again.
Now I noticed a new logentry.
Apparently one FAS-Request was stuck all the time. Now an access was logged that was served in 575s.

So I have to assume, that the call to ndsctl to determine the MAC was stuck and released after I restarted nodogsplash.

(I connect to the router via a VPN-tunnel from my central server to the router with OpenVPN. Over this tunnel I connect with SSH to get a console. This I currently do several times a day without problems. This only to complete the picture. ;-) )

**Comment by sqrt-1764** [comment]
Yesterday again: After about 4 days uptime of nodogsplash I noticed in my logs that the splash page was served slower as normal. When I then checked, ndsctl was stuck again.
This time no long redir-urls. Only 3 devices in the WLAN. Me (an Android Phone and a Linux Laptop and an iPhone - all not doing much in the network).

When I migrate to FAS Secure 2 soon, this problem might become obsolete for me, because then I get the MAC in the initial request and do not have to query ndsctl a second time any more.

So for me this is low priority - nevertheless interesting what is happening here.

**Comment by mwarning** [comment]
Am memory leak was discovered and fixed today. That may very well be the problem: https://github.com/nodogsplash/nodogsplash/commit/6b3b8b82d2ec7efb2634f6cefd0fbfe78cccf7f6

**Comment by bluewavenet** [comment]
@mwarning 
The memory leak is in the show_templated_page function and is not used when FAS is enabled so I don't think it is related to this issue.

**Comment by bluewavenet** [comment]
@mwarning @sqrt-1764 
Well, a bit of testing and I find **I can replicate the symptoms**.

A simple shell script that runs "ndsctl json" in a tight loop did the business. I actually left it running for over an hour as I went away to eat, but will re-run for measured times or even count cycles.
After stopping the test, "top" showed two instances of "ndsctl json" still running but using 0% cpu. Killing these with the "kill" command did not return things to normal.

Test was on a Gl-iNet AR750 (128MB ram, OpenWrt Ath79 Snapshot), with NDS v4.0.1.

The test script:
[test.sh.txt](https://github.com/nodogsplash/nodogsplash/files/3453606/test.sh.txt)

**Comment by bluewavenet** [comment]
Update: I left this running overnight and it locked up at ~ 500,000 iterations.
So a refinement of the script. It now not only counts iterations but also spawns each ndsctl command into the background.
Now it fails consistently at between 100 and 200 iterations. 
It seems ndsctl looses track if there are too many concurrent requests.
There are no free memory issues at the point of failure and nothing obvious in the logs at first sight.
[test2.sh.txt](https://github.com/nodogsplash/nodogsplash/files/3455180/test2.sh.txt)

**Comment by bluewavenet** [comment]
@sqrt-1764 
Does your FAS call ndsctl immediately, or does it do so only after FAS credential validation?

Some CPD implementations will poll the CP quite rapidly on detection, for example Firefox desktop does so every 3 seconds.

If your FAS calls ndsctl immediately this will mean it will repeatedly do so on each poll from the CPD.

My implementation of FAS only calls ndsctl AFTER the client credentials have passed verification and just before calling NDS for final authentication.

This could explain why I have not encountered the issue but you have.

We have nevertheless uncovered an issue somewhere in ndsctl I think.

**Comment by bluewavenet** [comment]
@mwarning 
Do you have an understanding of how ndsctl handles concurrent requests?

**Comment by sqrt-1764** [comment]
@bluewavenet 
Yes, currently my script calls ndsctl on every request. This in order to get the mac-address of the requesting device with which I query my backend-server for details for this user. (Identifying the user via his device ...)
So yes - this leads to more calls to ndsctl than other use cases.

With FAS Secure 2 I can circumvent this, but I hadn't time to implement this yet.

**Comment by bluewavenet** [comment]
@sqrt-1764 @mwarning 
We definitely have some sort of concurrent request problem with ndsctl and I will do some more investigation.
@sqrt-1764 
I have no idea what your FAS login page looks like, but for now you could mitigate the issue by not calling ndsctl until the client user clicks a "Continue" button. This will prevent every CPD trigger from causing an ndsctl call. CPD triggers can build up to a frequency of many times per second in a busy environment.

**Comment by sqrt-1764** [comment]
@bluewavenet 
There is shurely room for improvement at my FAS page. Currently we are still in development, so I can live with the current situation as the problem only rarely occurs at the moment.  
As FAS Secure 2 is the way to go, I will concentrate my energy there once I find the time. ;-)

Besides that:
I think you are right: I observed the problem when it seemed that two devices made quasi simultaneous requests to the router. Because of my "expensive" calls to ndsctl the problem now surfaced. But I think it occurs also otherwise - but maybe rarely. So it is good we found it. Now it can be fixed. ;-)

**Comment by mwarning** [comment]
@bluewavenet I can take a look. But it may take a few days until I have time to do so.

**Comment by bluewavenet** [comment]
@mwarning  @lynxis 
The problem might be in the area of the epoll queue.
ndsctl is fine with 2 concurrent calls, but with 3 the problem arises.
A refined script to demonstrate:
[test3.sh.txt](https://github.com/nodogsplash/nodogsplash/files/3463123/test3.sh.txt)

This starts 3 instances of ndsctl status concurrently. Then if you try to run another manually, it hangs and a restart of NDS is required to restore functionality. (it is the same with ndsctl json)

Editing the script and changing iterations to 2 and running and there is no problem (after an NDS restart).

This is an issue because to get the client token when running FAS at secure level 1, ndsctl is called. If two clients try to log in simultaneously we get a hang.

This is fairly rare but will happen sooner or later. Many client devices will poll CPD rapidly on connection and if the FAS immediately calls ndsctl rather than waiting for the client user to log in, the NDS call could be as often as once every 3 seconds with Firefox desktop and similar with other CPD implementations.

So a FAS not waiting for user interaction before calling ndsctl will encounter this problem much sooner, as in the case of @sqrt-1764 .

Of course, using  FAS secure level 2 completely avoids the issue as it does not need to call ndsctl..

**Comment by bluewavenet** [comment]
Lots more tests and it looks pretty much like it is a LOCK_CLIENT_LIST deadlock.
This causes ndsctl to lock up and also prevents any new clients from being set to preauthenticated.

**Comment by bluewavenet** [comment]
Fixed with pr #411 
Will have to think about another release :-D
